Add support for Flaresolverr and Jackett, fix UIDs, add some useful scripts, update README

Pavlogal opened the pull request:
I had some time on hand and I really like this project so I did some updates and cleaning up.

Changes:
- I noticed Jackett support had disappeared at some point so I reintroduced it, thus addressing: https://github.com/Luctia/ezarr/pull/10#issuecomment-1442821059
- added flaresolverr thus solving https://github.com/Luctia/ezarr/issues/48#issue-2213261081
- fixed audiobookshelf thus solving https://github.com/Luctia/ezarr/issues/44#issue-2129818996
- added automatic timezone detection to main.py (it just reads from /etc/timezone),
- UIDs were all over the place, unsorted, sometimes missing and mismatched. I fixed it all, but at the cost of some being different which could possibly cause trouble for old users re-running the script to get the new services. That's why I added `remove_old_users.sh` which removes the old users so that new ones can be created without conflicts. I'm considering making some even more drastic changes in the future to make UID's easily customisable instead of hardcoded everywhere but this is good enough for now.
- changed sample docker-compose to include all services with correct IDs, changed default root folder to just "/ezzar" which partly addresses https://github.com/Luctia/ezarr/issues/41#issue-2088993391
- added an update script
- added `write_docker_compose.py` which is essentially `main.py` but without the permission and folder creation
- updated README to reflect changes and added a note for NFS users to save them some trouble

Luctia commented:
Thanks for the work! I'll try to look at this tomorrow.

Ulimn commented:
I came to the GH issues because of the audiobookshelf and overseerr UID conflict. Glad you're already working on it. Appreciate the effort! :)

Ulimn commented on README.md:
```diff
@@ -43,20 +43,24 @@ crash*. It doesn't necessarily mean it will run well on your system ;) It featur
    (for example, running PleX and Jellyfin at the same time is a bit unusual), you can comment them
    out by placing `#` in front of the lines. This ensures they are ignored by Docker compose.
-6. Run `docker compose up`.
+   You also need to replace the timezones and root folders. The default timezone is Europe/Amsterdam
```
Didn't you change it to Sarajevo?

Pavlogal replied:
I did that on accident, good catch

Ulimn commented on README.md:
```diff
@@ -76,6 +80,13 @@ take a look at [important notes](#important-notes) before you continue.
+- NFS shares' permissions are mapped by user IDs. If you want to access a file as a client, your user ID needs to match the user ID of the owner (or group) of that file on the NFS server. 
+Note that if you are a group member (and not the owner), having matching group IDs won't be enough, there also needs to be a corresponding user on the NFS server. The easiest way to make sure
+the users and groups are set up on both sides correctly is to run setup.sh on both your NFS server and your client. Preferably run it on the server first.
+If you are running this script on the client make sure that you temporarily enable -no-root-squash on your NFS server, as the script needs superuser privileges to run and by deafult on NFS the root
```
Typo: default

Luctia reviewed (changes requested):
Again, thanks for the work! I've added a few comments, but there's also a more fundamental discussion to be had here. First off, I'm not sure why the re-ordering of user IDs was necessary, especially considering the potential compatibility issues if people use this script and already have permissions set on an existing library. I imagine they don't really want to remove all their old users and groups just to add the new ones in a different order, which (AFAIK) means they need to re-set their folder permissions). If you can make a compelling argument for it, I'm okay with it, but otherwise I prefer just using the old UIDs. Someone else actually mentions system users, and I suppose that _might_ be a nice change, although this, too, is mostly administrative.

The re-ordering of everything also makes it quite difficult to review the code. Not much to be done about this now, but wanted to mention it for PRs you might make in the future.

Pavlogal commented:
> Again, thanks for the work! I've added a few comments, but there's also a more fundamental discussion to be had here. First off, I'm not sure why the re-ordering of user IDs was necessary, especially considering the potential compatibility issues if people use this script and already have permissions set on an existing library. I imagine they don't really want to remove all their old users and groups just to add the new ones in a different order, which (AFAIK) means they need to re-set their folder permissions). If you can make a compelling argument for it, I'm okay with it, but otherwise I prefer just using the old UIDs. Someone else actually mentions system users, and I suppose that _might_ be a nice change, although this, too, is mostly administrative.
> 
> The re-ordering of everything also makes it quite difficult to review the code. Not much to be done about this now, but wanted to mention it for PRs you might make in the future.
This is my 2nd ever PR and first time using git 😅, so some of my commits are a mess and could've been better separated. As for the reordering I do recognize that I shouldn't make these sorts of decisions on my own. There were some conflicts like Audiobookshelf - Overseer and Bazarr - Jellyseer. Jackett that previously used 13008 had disappeared at some point. One of each pair inevitably had to be changed which would inevitably break something on someone's config. I figured I might as well reorder the whole thing so it goes in order of selection in `main.py`, (which ultimately makes it harder to review _now_, but easier later on) and then add a script for removing old users, which actually came in handy for me too. I **know** this is a janky and less than ideal solution but I figured it was better to push *some* solution until I do it properly later. The python files are altogether pretty repetitive and just a template of what each docker-compose service and useradd/permission should look like. Changing one ID requires changing both helper programs. I plan on possibly completely rewriting it down the line with some more elegant solution like letting the user globally choose their own IDs without editing each of the files. 
Of course I do see your side of the argument as well so I'll understand if you reject this change and I'll try to revert it.

> If you'd like to separate some of these changes (like the addition of the new services) from the changed UIDs while we discuss it, that might lead to that separate PR being merged sooner.
Will consider it, might do it tomorrow.

Pavlogal commented on write_docker_compose.py:
I agree 100%, this was just a quick stop-gap solution and I'll make something much better down the line

Pavlogal commented on docker-compose.yml:
Jackett was there previously but disappeared at some point, I just reintroduced it. It does work though. Flaresolverr is the new thing and it also works.

Pavlogal commented on docker-compose.yml:
```diff
@@ -8,10 +8,10 @@ services:
       - PGID=13000
       - UMASK=002
-      - TZ=Europe/Amsterdam
+      - TZ=Europe/Sarajevo
```
Accidental, I generated the file by running the compose builder file while pressing enter as a way of saying yes to everything. It also picked up my timezone from the automatic detection I added 😅

Pavlogal commented on README.md:
```diff
@@ -43,20 +43,24 @@ crash*. It doesn't necessarily mean it will run well on your system ;) It featur
-2. Copy `.env.sample` to a real `.env` by running `$ cp .env.sample .env`.
-3. Set the environment variables to your liking. Note that `ROOT_DIR` should be the directory you
-   have cloned this in.
+2. ~~Copy `.env.sample` to a real `.env` by running `$ cp .env.sample .env`.~~
+3. ~~Set the environment variables to your liking. Note that `ROOT_DIR` should be the directory you
+   have cloned this in.~~
```
I figured that this was a remainder of a previous implementation because docker-compose didn't use those variables anywhere, I should've asked you about that first

Ulimn commented:
> There were some conflicts like Audiobookshelf - Overseer and Bazarr - Jellyseer.

Can it be solved by introducing previously unused user ids? 2 of the above would have new ones (maybe Audiobookshelf for example as I assume it's not as widely used as Overseer). If I'm not mistaken, this would have the least impact.

Since they share the UID, one has to be changed, it cannot stay like this.

Pavlogal commented:
> > There were some conflicts like Audiobookshelf - Overseer and Bazarr - Jellyseer.
> 
> Can it be solved by introducing previously unused user ids? 2 of the above would have new ones (maybe Audiobookshelf for example as I assume it's not as widely used as Overseer). If I'm not mistaken, this would have the least impact.
> 
> Since they share the UID, one has to be changed, it cannot stay like this.

I am currently working on a new PR that addresses this, and all the other complaints. The order of IDs will be kept the same, and conflicts will be resolved as follows: Overseer will remain as is and Audiobookshelf will be moved to the first available ID of 13014. Jackett will use its old ID of 13008.

However I want to consult @Luctia on some things first, so that this can go as smoothly as possible. 
1. Is there any particular reason why every folder is owned by the user? I think it's much more elegant to just give each folder to their respective container and add the user to the mediacenter group instead of vice versa. I actually already changed this in this PR and it works well.
2. About that `.env`... I see it was used at some point but right now it does nothing as the provided `docker-compose.yml` is hardcoded, the only environment variable it uses is $UID. And on a fresh install docker actually throws a warning about it. That's why I crossed out those steps in README. Do we want to reintroduce this functionality (even though such file would be overwritten if someone were to run the python script) or just keep it as-is and remove the parts in README that mention it? Because...
3. docker-compose v1 vs v2... I spun up a new VM just for testing and currently this script works best on docker-compose v1 and v2 complains about an outdated `docker-compose.yml` version. V1 is probably still more popular but it's deprecated so we should at some point migrate to v2 and make it clear in README. I suppose this will require minor changes to how `docker-compose.yml` is structured but it brings some benefits as well, since the newer versions have extensions which can prove incredibly useful in this use case since they can do everything environment could, but more.

Luctia commented:
Great to hear you're working on a new version @Pavlogal! I was thinking about implementing some of your changes myself, but I don't want to invalidate your work.

>     1. Is there any particular reason why every folder is owned by the user? I think it's much more elegant to just give each folder to their respective container and add the user to the `mediacenter` group instead of vice versa. I actually already changed this in this PR and it works well.

I'm not sure what you mean by giving folders to their containers (this kinda already what is happening; a folder belongs to a user which is linked to the container). The structure of users and folders is based on the [servarr Docker guide](https://wiki.servarr.com/docker-guide#multiple-users-and-a-shared-group). They do mention an alternative which might be what you're doing, but I'd need to take another look.

>     2. About that `.env`... I see it was used at some point but right now it does nothing as the provided `docker-compose.yml` is hardcoded, the only environment variable it uses is $UID (for jellyfin and tautulli). Docker actually throws a warning about it since shell variables need to be exported first (and it's absent from the .env file), so in the end it can't find the variable and just throws a blank string as a default. That's why I crossed out those steps in README. Do we want to reintroduce this functionality? I can easily fix this and maybe I could even make a UID variable for each container so that users can choose their own IDs, thus giving them the power to use whatever order and range they like.

I actually saw that too when looking at your PR, an oversight on my end. I think the .env file is more elegant than changing it everywhere in the Compose file, but I think this should have no effect on the Python flow, just on the "take this compose file, comments some stuff out and throw it up" flow.

>     3. docker-compose v1 vs v2... currently both work but v2 complains about the version line in `docker-compose.yml`, since that feature has been deprecated and is now used just informatively for backwards compatibility. I think that can seamlessly be bumped up to 3.8 which is the latest version of the file spec that both v1 and v2 support, or it can be omitted completely if we want just v2 support.

If we're going to be changing versions anyway, I would prefer using the latest (stable) version. I think Compose v2 also works with older files, so that's fine, but the difference in command style (`docker-compose` vs `docker compose`) should be something we take into account when talking about scripting and documentation.

Good luck with the updated PR and feel free to discuss more! I'll try to answer quickly.

Pavlogal commented:
@Luctia thanks for the answer, really helped me understand what page we're on!

1. Anyway as for the 1st question I was referring to the python file for setting up users and groups. The python file actually does something differently to the script as there is a part that gives ownership of each folder to `$(id -u)`, aka the current user. I changed it to be more like the script so it gives each folder to their respective container and applies a 775 mask. I also added the user to the mediacenter group so that he can still have access to his files for any manual intervention. This change would finally synchronise what CLI and the script do and it should follow the recommended servarr config.

2. As for the env file, I already made a fix for that, I wrote a brand new docker-compose that utilizes all the env variables. It really is a more elegant solution. I suggest providing that file as a `docker-compose.yml.sample`, since a CLI generated file would otherwise override it and if someone wanted to switch to a manual installation they would have to git pull again.

3. Glad we're going with v2, seems to overall just work better. I adjusted the update script to reflect that but we should also make some notes in README so that people know that v2 is recommended.

I've been doing this on a new branch and I'll open a new PR tomorrow, though if you want to take an early peek you can check out my branch `oldorder`.
